snowfall/common.py

```diff
@@ -125,3 +127,115 @@ def save_training_info(
         f.write('best epoch: {}\n'.format(best_epoch))
 
     logging.info('write training info to {}'.format(filename))
+
+
+def read_lexicon(filename: str) -> Dict[str, List[List[str]]]:
+    '''Read a lexicon file.
+
+    Args:
+      filename:
+        Filename of the lexicon file. Every line in the file has
+        the following format::
+
+            word phone1 phone2 phone3 phoneN
+
+        For instance, an example file looks like the following::
+
+            HELLO HH AH0 L OW1
+            WORLD W ER1 L D
+
+        Columns are separated by space(s), tab(s), or both.
+        The first column is the word and the remaining columns
+        are the pronunciations of the word.
+
+        The second column can be pronunciation probabilities
+        of this word, which is discarded.
+
+    Returns:
+      Return a dict whose key is a word and the value is a list of
+      list-of-phones of the word.
+
+      For example, if the lexicon file contains::
+
+            HELLO HH AH0 L OW1
+            HELLO HH EH0 L OW1
+
+      The returned dict would be::
+
+        {'HELLO': [['HH', 'AH0', 'L', 'OW1'], ['HH', 'AH0', 'L', 'OW1']]}
+
+      Note:
+        We support words with multiple pronunciations.
+    '''
+    ans: Dict[str, List[List[str]]] = dict()
+    has_prob = None
+    with open(filename, 'r', encoding='latin-1') as f:
+        whitespace = re.compile('[ \t]+')
+        for line in f:
+            a = whitespace.split(line.strip(' \t\r\n'))
+            if len(a) < 2:
+                print(f'Found bad line: {line} in lexicon file {filename}')
+                sys.exit(1)
+            word = a[0]
+            phones = a[1:]
+            if has_prob is None:
+                try:
+                    p = float(phones[0])
+                    has_prob = True
+                except:
+                    has_prob = False
+            if has_prob:
+                phones = phones[1:]
+
+            if word not in ans:
+                ans[word] = [phones]
+            else:
+                ans[word].append(phones)
+    return ans
+
+
+def build_ctc_graph(lexicon: Dict[str, List[List[str]]], text: str,
+                    phone_symbol_table: SymbolTable,
+                    word_symbol_table: SymbolTable):
+    '''
+    Args:
+      lexicon:
+        It is returned by :func:`read_lexicon`.
+      text:
+        The transcript. It contains words separated by spaces.
+      phone_symbol_table:
+        Phone symbol table.
+      word_symbol_table:
+        Word symbol table.
+    '''
+    oov = '<UNK>'
+    assert oov in lexicon
+
+    prev_state = 0
+    whitespace = re.compile('[ \t]+')
+    words = whitespace.split(text.strip(' \t\r\n'))
+    rules = ''
+    for word in words:
+        if word not in lexicon:
+            word = oov
+
+        phones = lexicon[word]
+        phones = phones[0]  # TODO(fangjun): handle multiple pronunciations
```

Comment:
Maybe worse WER is because you force each word has only one pronunciation here for now?

Reply:
thanks, I will try it.

Reply:
You didn't also change the decoding graph to support the same repeated symbols?

Reply:
No, I reused the decoding graph from the existing code. Will change the training graph and decoding graph
in the way as you suggested.

Thanks!

Reply:
Fixed.